File: 207.course-schedule.java
// @leet start

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;
import java.util.Deque;

class Solution {
    Deque<Integer> schedule;

    public boolean canFinish(int numCourses, int[][] prerequisites) {
        this.schedule = new ArrayDeque<>();

        @SuppressWarnings("unchecked")
        List<Integer>[] adjList = new ArrayList[numCourses];

        for (int i = 0; i < numCourses; i++) {
            adjList[i] = new ArrayList<>();
        }

        for (int[] prereq : prerequisites) {
            adjList[prereq[1]].add(prereq[0]);
        }

        boolean[] visited = new boolean[numCourses];
        boolean[] recStack = new boolean[numCourses];

        for (int i = 0; i < numCourses; i++) {
            if (!visited[i]) {
                if (dfs(adjList, visited, recStack, i)) {
                    return false;
                }
            }
        }

        return true;
    }

    public boolean dfs(List<Integer>[] adjList, boolean[] visited, boolean[] recStack, int node) {
        if (recStack[node]) {
            return true;
        }
        if (visited[node]) {
            return false;
        }

        visited[node] = true;
        recStack[node] = true;

        for (Integer neighbor : adjList[node]) {
            if (dfs(adjList, visited, recStack, neighbor)) {
                return true;
            }
        }

        recStack[node] = false;
        schedule.addFirst(node);
        return false;
    }
}
// @leet end

File: 207.course-schedule.go
package leetcode

// @leet start
func canFinish(numCourses int, prerequisites [][]int) bool {
	// Kahn's algo for topo sort

	// Generate adjacency list
	adjList := make([][]int, numCourses)
	for i := range numCourses {
		adjList[i] = make([]int, 0)
	}

	// For calculating indegrees
	reverseAdj := make([][]int, numCourses)

	for _, prereq := range prerequisites {
		pre := prereq[1]
		post := prereq[0]

		adjList[pre] = append(adjList[pre], post)

		reverseAdj[post] = append(reverseAdj[post], pre)
	}

	// Compute indegrees
	zeros := make([]int, 0)
	indegrees := make([]int, numCourses)

	for i, fromNodes := range reverseAdj {
		indegrees[i] = len(fromNodes)
		if indegrees[i] == 0 {
			zeros = append(zeros, i)
		}
	}

	count := 0
	for len(zeros) > 0 {
		curr := zeros[0]
		zeros = zeros[1:]
		count++

		for _, node := range adjList[curr] {
			indegrees[node]--
			if indegrees[node] == 0 {
				zeros = append(zeros, node)
			}
		}
	}

	return count == numCourses
}

// @leet end

